proof/refine/ARM/Reply_R.thy

```diff
@@ -96,4 +95,487 @@ lemma replyRemoveTCB_st_tcb_at'_sym_ref:
   apply (clarsimp simp: obj_at'_def pred_tcb_at'_def)
   done
 
+lemma getReplyTCB_wp:
+  "\<lbrace>\<lambda>s. reply_at' rptr s \<longrightarrow> obj_at' (\<lambda>reply. P (replyTCB reply) s) rptr s\<rbrace>
+   getReplyTCB rptr
+   \<lbrace>P\<rbrace>"
+  unfolding getReplyTCB_def
+  apply (wpsimp)
+  apply (clarsimp simp: obj_at'_def)
+  done
+
+crunches setReplyTCB
+  for ksReadyQueues[wp]: "\<lambda>s. P (ksReadyQueues s)"
+  and ksSchedulerAction[wp]: "\<lambda>s. P (ksSchedulerAction s)"
+  and valid_queues[wp]: "valid_queues"
+
+crunches getReplyTCB
+  for inv: "P"
+
+lemma setReplyTCB_list_refs_of_replies':
+  "setReplyTCB tcb rptr \<lbrace>\<lambda>s. P (list_refs_of_replies' s)\<rbrace>"
+  unfolding setReplyTCB_def updateReply_def
+  apply wpsimp
+  apply (erule rsubst[where P=P])
+  apply (rule ext)
+  apply (clarsimp simp: list_refs_of_reply'_def obj_at'_def projectKO_eq
+                        list_refs_of_replies'_def opt_map_def
+                 split: option.splits)
+  done
+
+lemma setReply_valid_pde_mappings'[wp]:
+  "setReply p r \<lbrace>valid_pde_mappings'\<rbrace>"
+  unfolding valid_pde_mappings'_def setReply_def
+  apply (wpsimp wp: hoare_vcg_all_lift)
+  done
+
+lemma replyUnlink_valid_pspace'[wp]:
+  "replyUnlink rptr tptr \<lbrace>valid_pspace'\<rbrace>"
+  unfolding replyUnlink_def setReplyTCB_def getReplyTCB_def replyUnlink_assertion_def
+            updateReply_def
+  apply (wpsimp wp: sts'_valid_pspace'_inv hoare_vcg_imp_lift'
+              simp: valid_tcb_state'_def valid_pspace'_def)
+  apply normalise_obj_at'
+  apply (frule(1) reply_ko_at_valid_objs_valid_reply')
+  apply (clarsimp simp: valid_reply'_def)
+  done
+
+lemma replyUnlink_not_runnable_valid_queues:
```

Comment:
This should be unnecessary seeing that we have `replyUnlink_valid_queues` below

Reply:
Guh, good catch.